test/proof.js

```diff
@@ -37,30 +37,53 @@ tape('simple merkle proofs generation and verification', function (tester) {
           })
         })
       },
-      function (cb) {
+      function (cb) { // should create a valid proof of null
```

Comment:
Hm, I wish you had simply added new test cases instead of modifying existing ones. Are these 3 cases not valid anymore? Going through them I have a hard time distinguishing between different result of `verifyProof`:

- If proof is for a key that doesn't exist and `verifyProof` checks the same key, val and err are null
- If proof is for a key and `verifyProof` checks a totally random key (that wasn't in proof), there'll be error
- In the first test case, proof is for `key2bb`, then it tries to verifyProof for `key2`, and it again returns null for val and err. I.e. the same proof is also a valid proof of absence for `key2`, which is certainly interesting, although maybe unexpected?

I'm not sure if users would need to distinguish between all the various states? One benefit of the previous approach in my eyes is that only the value for which the proof was generated would be verified.

Reply:
> If proof is for a key that doesn't exist and verifyProof checks the same key, val and err are null

the proof should prove the key is associated with `null`. It returns the correct value -> `null`, and it does not raise an error, because the verify function, having been called on that key, returned the sufficient nodes to prove this.

> If proof is for a key and verifyProof checks a totally random key (that wasn't in proof), there'll be error

since this proof does not contain sufficient nodes to prove the random key, `verifyProof(randomKey)` will return an error. It does not matter if randomKey exists or not. It matter that the proof contains the correct nodes to verify such.

> In the first test case, proof is for key2bb, then it tries to verifyProof for key2, and it again returns null for val and err. I.e. the same proof is also a valid proof of absence for key2, which is certainly interesting, although maybe unexpected?

Yes, this was an interesting edge-case that I wanted to test. I did not properly document it. In this case, the proof _happens_ to contain enough nodes to prove the value at `key2` because traversing into `key22` would touch all the same nodes as traversing into `key2` (and an extra). So the proof from `key22` will be valid to prove the value of any shorter key.

> One benefit of the previous approach in my eyes is that only the value for which the proof was generated would be verified

I dont think thats a benefit except maybe for testing. It also has a major drawback in that it does not allow proofs to be efficiently combined. This approach allows the "proofTree" to be drop-in replacement for a "regular tree" (it can even do `put`). A light client could in the future process state transitions using existing EVM software.